gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/MongoDeviceRepository.java

```diff
@@ -69,19 +72,23 @@ public void init() {
     }
 
     @Override
-    public Flowable<Device> findByReferenceAndUser(ReferenceType referenceType, String referenceId, String user) {
-        var query = and(eq(FIELD_REFERENCE_ID, referenceId), eq(FIELD_REFERENCE_TYPE, referenceType.name()), eq(FIELD_USER_ID, user), gte(FIELD_EXPIRES_AT, new Date()));
+    public Flowable<Device> findByReferenceAndUser(ReferenceType referenceType, String referenceId, UserId user) {
+        var query = and(eq(FIELD_REFERENCE_ID, referenceId), eq(FIELD_REFERENCE_TYPE, referenceType.name()), userIdMatches(user), gte(FIELD_EXPIRES_AT, new Date()));
         var devicePublisher = rememberDeviceMongoCollection.find(query);
         return Flowable.fromPublisher(devicePublisher).map(this::convert);
 
     }
 
+    private Bson userIdMatches(UserId user) {
+        return or(eq(FIELD_USER_ID, user.id()), and(eq(FIELD_USER_EXTERNAL_ID, user.externalId()), eq(FIELD_USER_SOURCE, user.source())));
```

Comment:
**todo**: why are you managing the userIdMatches method in a different manner than for JDBC?

Here is what you did for JDBC
```
if (user.isExternal()) {
     return where(USER_ID_FIELD).is(user.id()).or(where(USER_EXTERNAL_ID_FIELD).is(user.externalId()).and(USER_SOURCE_FIELD).is(user.source()));
        } else {
            return where(USER_ID_FIELD).is(user.id());
        }
    }
```